core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java

```diff
@@ -2293,8 +2293,12 @@ private SqlNode registerFrom(
           alias,
           forceNullable);
 
-      // Keep preceding "LATERAL" keyword from joins with inner SELECT subquery
-      if (lateral && kind == SqlKind.SELECT) {
+      // Preserve "LATERAL" keyword in validated node:
+      // 1. Joins with inner SELECT subquery preceded by LATERAL
+      //    Example: SELECT ... FROM LATERAL(SELECT ...) AS t
+      // 2. Joins with UNNEST operator preceded by LATERAL
+      //    Example: SELECT ... FROM LATERAL(UNNEST ...) AS t
+      if (lateral && (kind == SqlKind.SELECT) || (kind == SqlKind.UNNEST)) {
```

Comment:
should this be `if (lateral && (kind == SqlKind.SELECT || kind == SqlKind.UNNEST))` to denote that LATERAL must be present? 
otherwise this could match `SELECT ... FROM (UNNEST ...) AS t` as well

Reply:
In this function (and Calcite in general), UNNEST will always imply LATERAL (i.e `lateral = true`), see [here](https://github.com/linkedin/linkedin-calcite/blob/li-1.21.0/core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java#L2273). So in theory we should never see `SELECT ... FROM (UNNEST ...) AS t` but I agree that the conditional doesn't read as nice so changed :)

Reply:
In this function (and Calcite in general), UNNEST will always imply LATERAL (i.e lateral = true), see [here](https://github.com/linkedin/linkedin-calcite/blob/li-1.21.0/core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java#L2273), so both conditionals are equivalent but I agree that this conditional doesn't read as nice so changed :)